Feature flags - config scripts refactor

Jlisowskyy opened the pull request:
Things done in this PR:
- reduced 8 (permutations of config flags) config sets from configure.bash script to only 2 (debug and release)
- introduced new way of configuring the kernel build via so called 'Feature flags', which allows to decide which of the feature is enabled in current build
- refactored code using new feature flags

kryczkal commented on scripts/config/configure.bash:
```diff
@@ -0,0 +1,340 @@
+
+  pretty_info "Creating cmake configuration files..."
+  # prepare conf.generated.cmake
+  local conf_cmake="${CONFIGURE_DIR}/../config/conf.generated.cmake"
```
This should be a function

kryczkal commented on scripts/config/configure.bash:
```diff
@@ -0,0 +1,340 @@
+}
+
+generate_cxx_feature_flag_files() {
+  # Source the feature flags to load existing flags
```
This function has high cognitive complexity, should probably be broken down into smaller parts.

kryczkal commented on scripts/config/configure.bash:
```diff
@@ -0,0 +1,340 @@
+CONFIGURE_BUILD_TYPE=""
+CONFIGURE_VERBOSE=false
+
+help() {
```
process args / parse args and help span a lot of the script and are repeated everywhere. This is not "DRY" at all in our code. Should probably be abstracted away to some interface like argparser.

Jlisowskyy replied:
We could do this in future, please add the issue to the github.

kryczkal reviewed (commented):
1. In my opinion feature_flags.conf should be feature_flags.yaml
same logic as .env_template and .env. This would make the configuration easier for user. Now having two files is a bit confusing. 
2. ".generated" files in config should be moved somewhere else

Jlisowskyy commented on alkos/kernel/arch/x86_64/common-loader-all/multiboot2/extensions.tpp:
```diff
@@ -10,8 +10,8 @@ namespace multiboot
-    const u32 kType      = TagNumber<Tag>::value;
-    const char *kTagName = TagNumber<Tag>::kTagName;
+    const u32 kType = TagNumber<Tag>::value;
+    // const char *kTagName = TagNumber<Tag>::kTagName;
```
GCC outputs a lot of warnings about unused variables, mostly because they were used in trace statements that you previously commented out, leaving the variables unused. Initially, I wanted to clean up the warnings, but eventually I gave up...

Jlisowskyy commented on alkos/kernel/arch/x86_64/common-loader-all/abi/debug_terminal.hpp:
```diff
@@ -3,23 +3,24 @@
 {
     /* verify if the usage is permitted */
-    R_ASSERT_TRUE(kUseDebugOutput);
+    R_ASSERT_TRUE(FeatureEnabled<FeatureFlag::kDebugOutput>);
```
refer to comment above

Jlisowskyy commented:
".generated" files in config was moved as mentioned previously in todos, resolved all todos slightly refactored the code and improved additionally by presets

Jlisowskyy commented:
Updated readme files and resolved other todos. Please as an answer to your first comment read the README.md in the config directory

Jlisowskyy commented:
Updated the config to also be stored in yaml. Example: 
```yaml
# Feature flags configuration - rerun configure script to update the build system
feature_flags:

  # Instead of proceeding to usual kernel boot enter the test mode, which allows to run single test on the kernel.
  run_test_mode: false

  # Enables additional debug information for spinlocks, which can help to identify deadlocks and other synchronization issues.
  debug_spinlock: true

  # Enables additional terminal output directed to serial port, mostly used for qemu builds and debugging.
  debug_output: true

  # Enables additional debug traces in the kernel, which can help to identify issues during execution.
  debug_traces: true
```

Renamed files tor feature_flags_config.yaml and feature_flags_schema.yaml

kryczkal reviewed (approved):
Readme is now outdated. Just delete it tbh. Also feature_flags_config.yaml is not being properly generated when i run either alkos_cli -c or ./scripts config configure so its broken. Also when I provide x86_64 to the script it issues a warning that arch was not found? but when you rerun it it's gone. Looks like this became a little convoluted.
